README.md

```diff
@@ -433,3 +434,334 @@ with the following lines.
+  makeScalaBlock scalaModule config
+```
+
+### Adding simulation options hooks
+Now we need to describe how to simulate our blocks.
+There are two main publish/subscribe targets for adding simulation option hooks:
+`dutSimCompileOptionsHooks` and `dutSimExecuteOptionsHooks` for compile and
+runtime options respectively. These hooks are functions of the type
+`DUT => Option (a => a)`. A `None` return value means that no modifications
+are made. Otherwise the returned transformation on `a` is applied.
+
+`dutSimCompileOptionsHooks` operate on `DUTSimCompileOptions`.
+`DUTSimCompileOptions` is defined as follows:
+```wake
+tuple DUTSimCompileOptions =
+  global IncludeDirs:    List String
+  global Defines:        List NamedArg
+  global SourceFiles:    List Path
+  global Plusargs:       List NamedArg
+```
+
+An example hook might look like:
+```wake
+publish dutSimCompileOptionsHooks =
+  def hook dut =
+    if dut.getDUTName ==~ "myDUT"
+    then
+      def optionsTransform options =
+        options
+        | editDUTSimCompileOptionsIncludeDirs ("myIncludes", _)
+        | editDUTSimCompileOptionsDefines (
+          NamedArg        "MY_DEFINE",
+          NamedArgInteger "MY_DEFINE_INTEGER" 1,
+          NamedArgDouble  "MY_DEFINE_DOUBLE"  1.0,
+          NamedArgString  "MY_DEFINE_STRING"  "string",
+          NamedArgPath    "MY_DEFINE_PATH"    "mySourceFile".source,
+          _
+        )
+        | editDUTSimCompileOptionsPlusargs (NamedArgString "my_plusarg" "value", _)
+        | editDUTSimCompileOptionsSourceFiles ("mySourceFile".source, _)
+      Some optionsTransform
+    else None
+  hook, Nil
+```
+
+The `pio` block needs `pio.sv` and `loopback.sv` to be added to the list
+of files to be compiled in the simulation so we need to add hooks to
+`dutSimCompileOptionsHooks` to add those files. There is a helper function
+`makeBlackBoxHook` that takes a name and a transform function and returns
+a hook that will apply the transform is the blackbox of the same name is found
+in the `DUT`. We can use this to tell the simulator about our pio/loopback
+source files because `duh` will generate blackboxes of those modules for us.
+Copy the following hook defninitions into `pio.wake`.
+```wake
+publish dutSimCompileOptionsHooks = pioHook, loopbackHook, Nil
+
+def loopbackHook =
+  def name = "loopback"
+  def addSources = source "{blockPIOSiFiveRoot}/rtl/loopback/loopback.sv", _
+  makeBlackBoxHook name (editDUTSimCompileOptionsSourceFiles addSources)
+
+def pioHook =
+  def name = "pio"
+  def addSources = source "{blockPIOSiFiveRoot}/rtl/pio/pio.sv", _
+  makeBlackBoxHook name (editDUTSimCompileOptionsSourceFiles addSources)
+```
+
+## Making a test
+
+Currently, only c integration tests are supported. A simple test is included
+with this repository in `block-pio-sifive/tests/demo/main.c`. It looks like this.
+```c
+int main()
+{
+  // read/write to axi block
+  volatile uint32_t * pio = (uint32_t *) 0x60000;
```

Comment:
This part we should update once @sistillson's stuff merges in, since we shouldn't be hardcoding the base address of the PIO anymore after that's in.